### PR TITLE
Add a vessel composite surface with ellipsoids on top and bottom.

### PR DESCRIPTION
**Pull request by nsedayilmaz**

### CODE REVIEW

File: openmc/model/surface_composite.py [hunk 1]
```diff
@@ -1837,3 +1837,49 @@ def __init__(self, center_base: Sequence[float], axis: Sequence[float],
...
+
+        """
+        Coefficients for quadric surface to create an ellipsoid
+
+            General equation for an ellipsoid: 
+                (x-xo)^2/r^2 + (y-yo)^2/r^2 + (z-zo)^2/h^2 = 1
+                
+            General form of a quadric surface equation:
+                Ax^2 + By^2 + Cz^2 + Gx + Hy + Jz + K = 0
+                """
```
Comment (pshriwise) [inline openmc/model/surface_composite.py:1865]:
```suggestion
        """
        General equation for an ellipsoid: 
        (x-xo)^2/r^2 + (y-yo)^2/r^2 + (z-zo)^2/h^2 = 1               
        """
```

Suggesting we reduce this comment to it's critical piece -- we cover the GQ equation in the documentation as it is.

File: openmc/model/surface_composite.py [hunk 2]
```diff
@@ -1837,3 +1837,49 @@ def __init__(self, center_base: Sequence[float], axis: Sequence[float],
 
     def __neg__(self) -> openmc.Region:
         return +self.plane_bottom & -self.plane_top & -self.cone
+
+class Z_Vessel(CompositeSurface):
```
Comment (pshriwise) [inline openmc/model/surface_composite.py:1841]:
```suggestion
class ZVessel(CompositeSurface):
```

File: openmc/model/surface_composite.py [hunk 3]
```diff
@@ -1837,3 +1837,49 @@ def __init__(self, center_base: Sequence[float], axis: Sequence[float],
...
+    """Vessel as a composite surface parallel to z-axis
+        includes composite surfaces with a ZCylinder and 
+        semi-ellipsoids on top and the bottom of this cylinder.
+        """
```
Comment (pshriwise) [inline openmc/model/surface_composite.py:1845]: We'll want to include some parameter descriptions in the documentation string here.

**Review by pshriwise — CHANGES_REQUESTED** [review]
Thanks @nsedayilmaz! This is a nice addition and great application for the composite surface. I've made a few small comments here. We'll want to add some tests for this surface as well.

**Review by paulromano — APPROVED** [review]
I did some refactoring on this branch to generalize it so it can be oriented in the x-, y-, or z-directions instead of just z. @pshriwise if you can take another quick look, I'll need your approval to merge. Thanks!

File: openmc/model/surface_composite.py [hunk 4]
```diff
@@ -1837,3 +1839,96 @@ def __init__(self, center_base: Sequence[float], axis: Sequence[float],
...
+        (x, y) basis. Defaults to (0,0).
+    axis : {'x', 'y', 'z'}
+        Central axis of the cylinder.
+
+    """
+
+    _surface_names = ('cyl', 'plane_bottom', 'plane_top', 'bottom', 'top')
+
+    def __init__(self, r: float, p1: float, p2: float, h1: float, h2: float,
+                 center: Sequence[float] = (0., 0.), axis='z', **kwargs):
```
Comment (pshriwise) [inline openmc/model/surface_composite.py:1876]: Should there be a type hint for `axis` here?

**Review by pshriwise — CHANGES_REQUESTED** [review]
Maybe a missing type hint but other than that it looks good to me!

**Review by pshriwise — APPROVED** [review]
Thanks @paulromano!

**Comment by nsedayilmaz** [comment]
Thank you both for working on this request!

**Pull request merged**